### ISSUE TITLE
Replace use of Java finalize()

### ISSUE THREAD

**Issue by ojwb** [swig/swig#1550]
In #1237 we discussed providing a way to avoid use of Java finalize() by using `PhantomReference` or `java.lang.ref.Cleaner` instead (the latter requires JDK 9, the former is supported since JDK 1.2).

I've started an attempt to implement this, so opening a ticket so that others know I am, and to have somewhere to track progress.

So far I have a hacked up version of the java typemaps such that `java.lang.ref.Cleaner` is used instead of generating a `finalize()` method.  Most of the testsuite passes, but from the failures it seems perhaps destruction isn't actually happening.

Once this is working, I'll try to adjust it to use `PhantomReference` instead.

**Comment by ojwb** [comment]
I fixed it so destruction actually happens, and mostly implemented this for the smart pointer wrappers too.  I'll push a WIP branch soon, once I've knocked off the test suite failures that I can see an easy fix for (currently about a dozen testcases fail).

I also tried the "create a lot of SWIG-wrapped objects in a loop" example, which with `finalize()` results in memory usage ballooning - the JVM runs `finalize()` with a low priority thread, which it seems doesn't get a look in if we're in a CPU intensive loop.  Even explicitly calling `obj.delete()` explicitly doesn't solve this - that means the C++ object deleted, but the Java object still exists.  Using a `Cleaner` (which uses `PhantomReference` behind the scenes) the objects get released as the loop runs - not instantly it seems, but in a timely enough fashion that the memory ballooning issue no longer seems to happen.

So that's the good news.

The bad news is I don't think we can transparently replace use of `finalize()` with a different mechanism.  E.g. I can't see how to use set a `javafinalize` typemap which instead creates the code I need without things breaking if a user sets a custom `javafinalize` typemap, because they won't know the right code to include in their `javafinalize`.

I also think I need to move `swigCPtr` and `swigCMemOwn` into a static inner class - I need something I can that can be accessed once the only reference to the proxy class is a `PhantomReference`.  I can't pull the values from the proxy class, because if I could there would be a real reference to it, and my code to do the deletion won't ever get called (this was the bug I fixed above).  That means that peeking and poking `swigCPtr` and `swigCMemOwn` directly no longer works, which seems to be widely assumed to be OK at least within the test suite.

Also something that I'm confused by is that currently a derived java proxy class has a `swigCPtr` member and so does the base class, and both appear to have the same value.  Currently I've eliminated the one in the derived class and if that's broken stuff I haven't spotted it yet.  Is there a reason for the current duplication I haven't spotted?

**Comment by ojwb** [comment]
> Also something that I'm confused by is that currently a derived java proxy class has a `swigCPtr` member and so does the base class, and both appear to have the same value. Currently I've eliminated the one in the derived class and if that's broken stuff I haven't spotted it yet. Is there a reason for the current duplication I haven't spotted?

I've spotted the problem here - when multiple inheritance is involved, the `swigCPtr` values aren't actually all the same.  Will try to address that then.

**Comment by ojwb** [comment]
I've opened PR #1554 with a WIP version of my changes so people (especially @wsfulton) can take a look.

These four tests currently fail - I've noted any analysis I've done as to why:

* java_throws: `./NoExceptTest.java:23: error: unreported exception MyException; must be caught or declared to be thrown` - testcase adds special handling via a modified `javafinalize` typemap which needs reworking for the new approach.
* java_typemaps_proxy: testcase replaces some of my new typemaps with modified versions of the current typemaps, and the mismatched combination is never going to work - I guess this testcase needs updating for the new typemaps, or we need a way it can accomplish what it wants without having to replace typemaps.
* default_constructor: `Exception in thread "main" java.lang.RuntimeException: Protected destructor exception should have been thrown` - not investigated yet
* li_boost_shared_ptr: `GC failed to run (li_boost_shared_ptr). Expected count: 1 Actual count: 44` - not sure what's going on here.

**Comment by ojwb** [comment]
@wsfulton What are your thoughts on this?

In #1237 you said:
> For anyone looking for a challenge, I'd really like to see some proposed alternative solutions using PhantomReference.

So I've been trying to implement this in SWIG itself, rather than just adding custom typemaps for Xapian (which would solve my immediate problem but not benefit other users of SWIG's Java backend).

I've not managed to get this to work with `PhantomReference` instead of `Cleaner` yet - I have an attempt to do so but the `PhantomReference` never seems to get added to the `ReferenceQueue`.  I can keep working on this (or push those changes if you or someone else with more Java knowledge wants to take a look) but if I'm only aiming to address this in Xapian then I can probably just stick to `Cleaner` and say JDK >= 9 is required.

**Comment by fflexo** [comment]
It should be possible to implement `Cleaner` using nothing but `PhantomReference` right? So one option might be to bring our own `Cleaner` in a fragment that's API compatible with the JDK >= 9 case but works back to to 1.2 and could easily be switched over later.

**Comment by ojwb** [comment]
> It should be possible to implement `Cleaner` using nothing but `PhantomReference` right?

I think I read something which suggested it was behind the scenes, so probably.

> So one option might be to bring our own `Cleaner` in a fragment that's API compatible with the JDK >= 9 case but works back to to 1.2 and could easily be switched over later.

I was trying a direct `PhantomReference` solution which can be a bit simpler because the situation is more constrained, but implementing a drop-in `Cleaner` replacement seems a sensible approach too.

However, one month on I've had zero feedback on this.  I really need to get this fixed in Xapian's bindings so I'm reluctantly giving up on trying to sort this out in SWIG.  I'd certainly much rather fix it for everyone, but I can't wait indefinitely.

So if anyone wants to take on this please do.  I'm happy to also push the work-in-progress `PhantomReference` attempt if that would be useful, and can try to answer questions about the changes.

**Comment by vadz** [comment]
Sorry for the lack of feedback. I'm very much interested in having this, but I don't know much about this stuff, so I just hoped to profit from the fruits of your labour once it would have been done...

Putting the problems with `PhantomReference` aside, what is the status of `java.lang.ref.Cleaner`-based version? I think with more and more time passing, requiring Java 9 may stop being a problem relatively soon, so maybe it would actually be enough?

In any case, let me reiterate that I very much appreciate your work on this, thanks a lot!

**Comment by ojwb** [comment]
> what is the status of `java.lang.ref.Cleaner`-based version?

Essentially fully working.

The whole testsuite passes on the branch with only a few minor tweaks to testcases, with the exception of two testcases.  The reasons behind those remaining failures are indicative of the problems that still need resolving, and they're really down to the how the current typemaps work.

The awkward thing is there doesn't really seem to be a clean division between what's part of SWIG and what's customisation in SWIG's Java-specific typemaps - for example, SWIG's `java.i` defines a `javafinalize` typemap, but the user can override this.  But if they do it completely replaces the standard one, and it seems if they want to just add to the standard behaviour they're expected to just copy the code from the standard typemap into their interface file.  That doesn't play well with my changes, because I define an empty `javafinalize` typemap as standard, so when the user defines one which includes a copy of the `javafinalize` from current SWIG things don't work.

Also, even if they want to replace the default `javafinalize` typemap, they are probably assuming that by doing so they will prevent the automatic cleaning up, which isn't the case with the Cleaner-based approach (because that's done elsewhere and the default `javafinalize` typemap is now empty).  If there were a "javafinalize"/"nojavafinalize" feature flag which toggled this on/off I could hook into that, but the current approach seems to be built a bit too closely around the assumption that the cleanup actually happens via a Java finalizer.

I'm also unsure how much of what the testsuite does with these typemaps is stuff we aim to support in user interface files, and how much is essentially poking around with internals for the purposes of implementing testcases.  If something is in the former category then it's important to try to support it, but for the latter category just changing the testcase is OK.

The timing relative to SWIG 4.0.0 also perhaps unfortunate, but I'm looking at this because a client wants it addressed in Xapian and they didn't ask until after SWIG 4.0.0 was released.

> I think with more and more time passing, requiring Java 9 may stop being a problem relatively soon, so maybe it would actually be enough?

I wouldn't fixate on my `PhantomReference` problems as I'm pretty sure that could be made to work without much difficulty.  I just stopped investigating further for now as it's a waste of my time if this isn't going to get merged into SWIG - requiring Java 9 is not a problem for Xapian or for my client given the huge improvements to memory management the change brings.

I'm currently working on implementing this in Xapian's interface files, which will at least provide a worked example for others who want to avoid use of `finalize()` in their own Java bindings.

**Comment by ojwb** [comment]
> I'm currently working on implementing this in Xapian's interface files, which will at least provide a worked example for others who want to avoid use of `finalize()` in their own Java bindings.

I now have a working version of this:

https://github.com/xapian/xapian/compare/master...ojwb:java-cleaner

Currently I have to post-process SWIG's output for three things:

* `java.cxx` hard codes `swigCPtr` which I need to rewrite to `swigWrap.swigCPtr`
* I end up getting `()` for `$jnicall` in some cases where it ought to be empty - not sure what's going on there
* Stripping out the constructors, etc from a couple of classes which just exist to house static functions because Java doesn't allow functions outside of any class - with my custom typemaps these produce code which fails to compile

The first I can't see how to avoid without modifying SWIG.  The other two might be solvable.

**Comment by wsfulton** [comment]
@ojwb, this is great work. Sorry, I havn't been tracking this, I'm not prioritising large disruptive changes like this right now, I feel I've been doing that for too long in getting SWIG 4 out the door! As you say the timing is unfortunate, as it's missed a major version change. I can't see any option other than putting this into a holding branch or something for version 4.1. Making fundamental changes to typemaps is really horrible for users that use customise typemaps. I have a round of these sort of changes to improve director exception handling which I abandoned for 4.0. I might take a fresh look at them and bring them back for 4.1. We need to implement a fail-safe way of detecting custom typemaps by changing the names of the typemaps or some clever inspection of the contents and then error out if they need upgrading. This is what I tried to do for the IDisposable changes in the C# destruction code in typemaps, see #421. Perhaps you can take a detailed look at these changes and take some inspiration from this as the area and nature of change is similar.

More generally though, I know a large company who is upgrading to a combination of Windows 10 and jdk8 next year and this will be the standard platform for the following 5 years. So, jdk9 only changes are not a good idea, so I suggest the `Cleaner` drop in replacement class idea is worth pursuing. Perhaps the drop-in replacement can be switched in/out easily by using the same class name but in a different package? Also more generally any work that is done and is battle hardened in Xapian will be good for inclusion into SWIG.

I will come back after looking at the nitty gritty details.

**Comment by kevinherron** [comment]
Is this still being considered for a future release? I think many people are on JDK 11 now and even many of those who aren't will be making the upgrade to JDK 17 instead, along with the JDK 11 folks. 

I don't see any mention of actually removing `finalize` yet but it makes me nervous.

**Comment by ojwb** [comment]
It seems something that really ought to be sorted out in SWIG, but I'm afraid I've pretty much given up on trying to do that myself.

**Comment by fflexo** [comment]
My analysis from looking at this previously was that the big problem arises from backwards compatibility issues. It's basically impossible to allow everything that worked before to work identically with Cleaner/PhantomReferences automatically. 

My best guess is that an approach similar to the one taken for Python 3 may be the best approach, i.e. gate it all behind a off by default command line flag, flip the default at some later major release once things have settled down enough? Maybe throw in a warning or two about relying on deprecated behaviour along the way?

**Comment by nioncode** [comment]
This sounds like a sensible approach. Then users that want to use the new functionality can opt-in early and then the new functionality will be made the default in the next (or even later) major version.

**Comment by tomwhoiscontrary** [comment]
Two random thoughts on this from a random user.

On backwards compatibility, i suspect the great majority of code does not use a javafinalize typemap. This code should carry on working as before, right? Whereas code that does use a javafinalize typemap is going to be making a load of assumptions which just don't hold in a PhantomReference-based approach, and will need to be rewritten to be correct. So i think explicitly breaking custom javafinalize is actually what you want here - the work it forces on some users is unavoidable. But then it's easy for me to say that, i won't be responding to the distraught bug reports!

Reworking destruction is also a chance to re-examine how it interacts with threads. I am using a C++ library which makes extensive use of the observable pattern; objects have normal pointers to objects they depend on, and vectors of pointers to objects which depend on them. Each object's destructor removes the dying object from its dependencies' dependent vectors. This is very much not thread-safe! But in the current SWIG implementation, where destructors are called on the finalizer thread (?), it is inevitably multithreaded. This is quite a headache. [The SWIG documentation does warn about this](https://www.swig.org/Doc1.3/Java.html#java_multithread_libraries), but only from the point of view of allocator thread-safety; i'm not sure the authors were aware of possible library-level thread-safety issues as well. What i would absolutely love is if SWIG could handle destruction in such a way that deletion did not occur on a GC thread, or some other thread that isn't my application thread. Or at least, if it gave me some sort of way to achieve this. Ideally, without having to put a custom destruction-related type mapping on every class.

The most luxurious option i can think of would be for SWIG to buffer pointers to C++ objects which are to be deleted, then delete everything in the buffer every time an application thread crosses into native code. Then, nobody would ever need to think about thread-safety of object deletion, it would just work - as long as you call wrapped methods now and then . You would probably also want a way to provoke explicit deletion of buffered objects, without calling a method in the wrapped library.

If that's too intrusive, it would still be a huge help to have some option to customise deletion behaviour globally.

**Comment by fflexo** [comment]
I don't think the "luxurious" option completely solves the threading problem because you can still have multiple threads entering and exiting from JNI calls and the semantics for which thread runs the destructor would still be complex to get right.

**Comment by ojwb** [comment]
Java seems to have escalated the deprecation warning to a "removal" warning and the suppression added by #1240 is no longer effective:

```
warning: [removal] finalize() in Object has been deprecated and marked for removal
```

I don't know when this started as I don't run the java tests often, but I'm seeing it with:

```
 java -version
openjdk version "18.0.2-ea" 2022-07-19
OpenJDK Runtime Environment (build 18.0.2-ea+9-Debian-2)
OpenJDK 64-Bit Server VM (build 18.0.2-ea+9-Debian-2, mixed mode, sharing)
```

> I will come back after looking at the nitty gritty details.

@wsfulton Did you manage to do that?

I'd mostly given up hope that this patch was going to go anywhere, but if Java is actually going to remove `finalizer()` SWIG will have to change to stop using it, and ideally before Java actually removes it...

**Comment by ojwb** [comment]
It seems SWIG is really going to need to stop relying on `finalize()` soon as it's now slated for removal from Java.

Hopefully the work I did on this that's in PR #1255 will be of some use, but it's been left to bit-rot for long enough that it'll need updating just to apply to current git so I've closed the PR as we have this issue.

**Comment by ojwb** [comment]
Sorry, wrong PR number - that should be #1554 (I'll edit the comment too).

**Comment by igorski** [comment]
Is there any "official word" on this matter ? I have been reading through the thread and am wondering what the current implementation requirements are for usage of SWIG within a Java context.

Previously, breaking reference to objects and having them garbage collected would invoke native layer destructors nicely through `finalize()`, but for reasons stated this is deprecated and using `Cleaner` might be an option instead. Question is : is/will this be transparently handled through SWIG or should consumer of the SWIG wrapped content explicitly call the cleaner API ?

**Comment by ojwb** [comment]
#2556 suppressed the removal warning mentioned above:

```
warning: [removal] finalize() in Object has been deprecated and marked for removal
```

However `make check-java-test-suite` now (or still?) emits multiple warnings such as:

```
checking java testcase director_pass_by_value (with run test)
./director_pass_by_value_runme.java:17: warning: [removal] runFinalization() in System has been deprecated and marked for removal
    System.runFinalization();
          ^
1 warning
```

I'm using OpenJDK "22-ea" apparently.

> Is there any "official word" on this matter ? I have been reading through the thread and am wondering what the current implementation requirements are for usage of SWIG wrapped libraries within a Java context.

The issue is still blocked on @wsfulton who said "I will come back after looking at the nitty gritty details" and hasn't yet (not trying to sound critical, just the reality is that's the blocker here).

I've sadly pretty much given up on progressing this (partly because my original work on this was funded by a client who since decided to take a different approach, but also because it now seems an unproductive use of my limited personal hacking time).

The key issue I think needs resolving is how to manage the impact on existing wrappers, because quite a lot of the innards are unfortunately exposed and might be overridden so it's hard to not break some existing wrappers with the change from finalizers to cleaners.  Perhaps the answer is we'll just have to break things because it seems finalizers are going away completely relatively soon (you now get a "removal" warning, not just a "deprecation" one).

> Previously, breaking reference to objects and having them garbage collected would invoke native layer destructors nicely through finalize(), but for reasons stated this is deprecated and using Cleaner might be an option instead.
>
> My question is : is/will this be transparently handled through SWIG (meaning breaking reference is enough to ensure that a native layer object will clean up after itself eventually) or should consumers of the SWIG wrapped content explicitly call the cleaner API ?

Once this gets resolved I would expect it to work similarly to currently (just be implemented with a cleaner instead of a finalizer method).  Hard to say for sure, but that's what I was aiming for.

If it doesn't get resolved before the Java developers actually remove finalizers entirely then I guess users of the SWIG-wrapped API will need to more manually manage object lifetimes, at least until it does get resolved.

**Comment by igorski** [comment]
Thanks for your reply and elaboration. It will likely become more urgent when the finalizer is fully removed as otherwise using SWIG with Java will be a little restrictive. Ideally the C-based library I'm wrapping for Java works with ancient versions (with finalizers) up to the latest JDK, but I'm happy to up the JDK requirements, as long as there is a universal solution.

Guess nothing to do but wait. In the meantime I did change the wrapped library API to provide `dispose()` methods when objects are no longer needed, while now the finalizer and GC invokes the destructor, in the future I can change this to use cleaner when needed without making breaking changes to implementations (of the SWIG wrapped library). It does require the user of the library to be more conscious about resource allocation which one hand is good, but on the other I liked the "worry not, this language has a garbage collector" approach...

**Comment by trim21** [comment]
the memory leak can be avoid by manully close the object with https://github.com/swig/swig/issues/1791

can we consider add option to disable the finalize method first?
